### PR TITLE
fix sampling and padding bugs and some small env issues

### PR DESCRIPTION
**Pull request by wenzhengzeng**
# Pull Request Description

I would like to thank all the authors for their excellent work and for open-sourcing this project to the community!

This pull request fixes several bugs in the current implementation (see below).

---

### What went wrong

When **`use_frame_steps=true`** (default setting of RoboMeter and RoboReward), multiple positions are sampled evenly from the full video to compute metrics. The previous implementation used the `enumerate` index as the retrieval index when aggregating predictions and targets (see the loop starting at [this line](https://github.com/robometer/robometer/blob/main/robometer/evals/compile_results.py#L363C13-L363C50), where **`timestep`** is the loop position, not the **actual subsampled frame index** in the original video). That results in the sampled indices being fixed to **`0, 1, 2, …`** (i.e. `range(0, custom_eval.subsample_n_frames)`, e.g. **`0 … 4`** when `subsample_n_frames=5`) for every video, instead of the **actual subsampled frame indices** in the original video (e.g. evenly spaced timesteps along the clip).

### What this PR does

I added **`original_sampled_index`** to record the sampled index in the original video. It is populated consistently (including after subsampling / padding), used in **`compile_results`** when aggregating metrics, and the `enumerate`-based path is kept only as a fallback when no valid index is available—so alignment follows the real sampling, not the loop position.

### Performance comparison (VOC average Pearson · policy ranking τ) based on my reproduction

| Metric (split)                      | ROBOMETER w/ RoboReward Training Data (from paper) | ROBOMETER (w/ RBM-1M) (from paper) | reimplementation before this fix | new implementation |
| ----------------------------------- | -------------------------------------------------- | ---------------------------------- | -------------------------------- | ------------------ |
| **VOC r** (RBM-EVAL-ID)             | **0.842**                                          | **0.921**                          | **0.841**                        | **0.860**          |
| **VOC r** (RBM-EVAL-OOD)            | **0.927**                                          | **0.948**                          | **0.927**                        | **0.886**          |
| **policy ranking τ** (RBM-EVAL-OOD) | **0.552**                                          | **0.655**                          | **0.649**                        | **0.649**          |

While reproducing the released code, I noticed that **before this fix**, VOC numbers line up more closely with the paper’s **“w/ RoboReward Training Data”** column than with the other reported columns. Could the authors help clarify whether this is expected, or whether some eval/config details on my side may still be misaligned with the paper?

---

### What went wrong

For **reward alignment** evaluation, the current behavior **includes padded frames** in the metric computation.

### What this PR does

To exclude padding, set `custom_eval.pad_frames=false` in the run script (If I understand correctly, the original Robo Dopamine setup did not use padding, and whether use padding will not influence the inference result, but it can change calculated metrics when padded repetitive frames are included in the metric aggregation).

---

## 3. Minor environment/requirements updates

I found that some dependency versions released after March 21 can introduce minor incompatibilities. This PR changes the requirements slightly so that packages do not conflict with each other.

---

## 4. vLLM GPU selection at startup (for Robo-Dopamine)

The previous startup path could overwrite externally provided GPU-related environment variables (e.g., `CUDA_VISIBLE_DEVICES` / `LOCAL_RANK`) with hardcoded defaults.

This PR changes the initialization logic to respect pre-set environment variables and only apply defaults when they are missing. This makes vLLM startup GPU assignment configurable from the launch environment (for example, shell scripts or scheduler settings), instead of being forced to a fixed device.

---

Thanks again for the solid work—looking forward to your feedback on this PR. :)

### CODE REVIEW

**Comment by jesbu1** [comment]
Re: the sample index issue. Did you for sure see an issue here? 

I don't believe this is an actual issue because `timestep` should always be equal to the length of `pred_array` and `tgt_array` since they come from the same `enumerate()` function. 

Therefore, indexing into pred_array via `timestep` is correct (but yes, the naming is confusing).

**Comment by aliang8** [comment]
Thanks for the PR and for digging into this.

On the `use_frame_steps=True` sampling/indexing point: I do not think this affects the offline benchmark results we report. The aggregation path that uses `timestep` is only entered when `custom_eval.use_frame_steps=true`. For the paper-style offline Robometer evaluation, as clarified in https://github.com/robometer/robometer/issues/22, we use `custom_eval.use_frame_steps=false`; in that mode each trajectory produces one result, and the metric code consumes the full prediction/target arrays directly rather than doing frame-step aggregation.

For the `use_frame_steps=True` path itself, the predictions and targets being indexed are already the sampler output after frame selection/subsampling. So `timestep` is not indexing into the raw original video; it is indexing into the already-sampled `pred_array` and `tgt_array`, which are aligned with each other. The variable name is admittedly confusing. 

On the RoboDopamine padding point: `pad_frames` only has an effect when the selected trajectory has fewer frames than the sampler's `max_frames`; it pads by repeating the final frame and the final progress target. For the OOD eval split we are using, the sampled trajectories are already 32 frames, so padding does not take effect. In any case, even if padding were applied, it repeats the terminal observation and terminal progress target, so the prediction/target alignment remains consistent and the metrics should still reflect the progress correctly rather than introducing an indexing mismatch.

We have also since updated the relevant evaluation results and reproduced them on our end; see the discussion in https://github.com/robometer/robometer/issues/22 for the updated command and numbers.

**Comment by aliang8** [comment]
Closing PR for now, feel free to reopen this did not address your concerns!

**Comment by wenzhengzeng** [comment]
@aliang8 @jesbu1 Thank you for taking the time to look into this issue and for the meaningful discussion!


###  About use_frame_steps

I agree that when `use_frame_steps=False`, the current code is correct.

However, I still think there is a remaining issue when `use_frame_steps=True`, especially when `subsample_n_frames` is set. I may still be misunderstanding the intended behavior, so please correct me if any part of the reasoning below is wrong.


I understand that `pred_array` and `tgt_array` are aligned with each other. My point is that the final aggregated `traj_preds` and ` traj_targets` used for metric computation **do not appear to be the arrays we actually want**.

A quick way to verify this is to inspect the final `traj_targets` in `robometer/evals/compile_results.py` (e.g., [here](https://github.com/robometer/robometer/blob/847353484eb3b46a74936e9b8932630b7378436e/robometer/evals/compile_results.py#L396) or [here](https://github.com/robometer/robometer/blob/847353484eb3b46a74936e9b8932630b7378436e/robometer/evals/compile_results.py#L461) when `use_frame_steps=True` and `subsample_n_frames=5`. The values do not span the trajectory in the expected way:

image-1

It can be observed that the final aggregated target trajectory tends to concentrate on earlier frames, rather than being uniformly distributed and eventually ending at 1. If my understanding is correct, `traj_targets` should cover the uniformly sampled trajectory endpoints and should end at `1`, like this:

image-4


Below is a concrete example showing why this happens.


## Concrete Example

The sample I used for debugging is:

```text
jesbu1_racer_rfm_racer_val/frames/trajectory_9ebf31b8-36f2-4595-911f-7c1385e5c7d4
```

The original `.npz` file for this sample contains 12 frames, indexed from `0` to `11`.

With the following settings:

```text
use_frame_steps=True
subsample_n_frames=5
```

the reward alignment sampler in `robometer/data/samplers/eval/reward_alignment.py` first samples 5 endpoint positions from the 12-frame trajectory. Conceptually, this gives endpoint frames:

```text
[0, 2, 5, 8, 11]
```

The sampler then creates raw prefix subsequences ending at those frames:

```text
[0]
[0, 1, 2]
[0, 1, 2, 3, 4, 5]
[0, 1, 2, 3, 4, 5, 6, 7, 8]
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
```

Then, during `__getitem__`, `_get_traj_from_data` enforces `max_frames` before returning the `Trajectory` object. In my setup, `max_frames=8`.

Therefore:

- if the prefix has fewer than 8 frames, it is right-padded by repeating the last frame and the last target value;
- if the prefix has more than 8 frames, it is uniformly subsampled down to 8 frames.

So the actual frame positions represented by each returned `Trajectory` are approximately:

```text
raw prefix [0]           -> [0, 0, 0, 0, 0, 0, 0, 0]          # padded
raw prefix [0, ..., 2]   -> [0, 1, 2, 2, 2, 2, 2, 2]          # padded
raw prefix [0, ..., 5]   -> [0, 1, 2, 3, 4, 5, 5, 5]          # padded
raw prefix [0, ..., 8]   -> [0, 1, 2, 3, 5, 6, 7, 8]          # subsampled
raw prefix [0, ..., 11]  -> [0, 2, 3, 5, 6, 8, 9, 11]         # subsampled
```


My understanding is that each prefix subsequence is meant to represent the model prediction at the endpoint of that subsequence. Therefore, when computing the reward alignment metrics, the aggregated trajectory should correspond to the endpoint frames:

```text
[0, 2, 5, 8, 11]
```

This would produce a curve that is uniformly sampled over the original trajectory, from the first frame to the final frame.

## Current Behavior

In [here](https://github.com/robometer/robometer/blob/847353484eb3b46a74936e9b8932630b7378436e/robometer/evals/compile_results.py#L363), when `use_frame_steps=True`, `last_frame_only=False`, the code falls back to the `enumerate` index. For the five subsequences above, `timestep` is:

```text
[0, 1, 2, 3, 4]
```


With the `max_frames=8` preprocessing above, those ordinal selections correspond approximately to original frames:

```text
[0, 1, 2, 3, 6]
```

instead of the intended endpoint frames:

```text
[0, 2, 5, 8, 11]
```

Although `pred_array` and `tgt_array` are internally aligned within each individual returned `Trajectory`, the reconstructed trajectory used for metric computation is not aligned with the intended uniformly sampled endpoints. It instead takes the `timestep`-th element from each increasingly long prefix, after padding or subsampling has already been applied, which biases the aggregated curve toward earlier frames.




### About padding in Robo-Dopamine

I understand the current implementation, and I agree that it is a reasonable choice. My thinking is that including padded frames in the metric computation may slightly bias the metric, since repeated padding frames effectively increase the weight of the final real frame. For example, in MSE, the final-frame error can contribute more heavily to the overall score because of padding. For that reason, I wonder whether excluding padded positions from the metric computation might be a slightly cleaner choice. That said, I do not think the current approach is necessarily wrong; I just wanted to flag this potential trade-off for clarity. :)


Thank you again for looking into these potential issues. I really appreciate the helpful discussion!
